staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -986,6 +1032,61 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	return allErrs
 }
 
+func extractMaxElements(schema *apiextensions.JSONSchemaProps) *uint64 {
+	switch schema.Type {
+	case "object":
+		if schema.AdditionalProperties != nil {
+			if schema.MaxProperties != nil {
+				maxProps := uint64(*schema.MaxProperties)
+				return &maxProps
+			}
+			return nil
+		}
+		// return 1, and not nil, because even if we're
+		// not looking at a map here, we can assume the object still preserves
+		// the cardinality of its parents
+		return uint64ptr(1)
+	case "array":
+		if schema.MaxItems != nil {
+			maxItems := uint64(*schema.MaxItems)
+			return &maxItems
+		}
+		return nil
+	default:
+		return uint64ptr(1)
+	}
+}
+
+// multiplyWithOverflowGuard returns the product of baseCost and cardinality unless that product
+// would exceed math.MaxUint, in which case math.MaxUint is returned.
+func multiplyWithOverflowGuard(baseCost, cardinality uint64) uint64 {
+	if baseCost == 0 {
+		// an empty rule can return 0, so guard for that here
+		return 0
+	} else if math.MaxUint/baseCost < cardinality {
+		return math.MaxUint
+	}
+	return baseCost * cardinality
+}
+
+func getExpressionCost(cr cel.CompilationResult, cardinalityCost costInfo) uint64 {
+	if cardinalityCost.MaxCardinality != nil {
+		return multiplyWithOverflowGuard(cr.MaxCost, *cardinalityCost.MaxCardinality)
+	}
+	return multiplyWithOverflowGuard(cr.MaxCost, cr.MaxCardinality)
+}
+
+func getCostErrorMessage(expressionCost, costLimit uint64) string {
+	exceedFactor := float64(expressionCost) / float64(ExpressionCostLimit)
+	if exceedFactor > 100.0 {
+		// if exceedFactor is greater than 2 orders of magnitude, the rule is likely O(n^2) or worse
+		// and will probably never validate without some set limits
+		// also in such cases the cost estimation is generally large enough to not add any value
+		return fmt.Sprintf("CEL rule exceeded budget by greater than a factor of 100 (try adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are used)")
+	}
+	return fmt.Sprintf("CEL rule of estimated cost %d exceeded budget of %d by factor of %.1fx (try adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are used)", expressionCost, costLimit, exceedFactor)
```

Comment:
I'm not sure presenting unitless CEL cost budget numbers to users is ever going to help them... I'd probably stick to the exceeded by X% or a factor of ##X style messages

Reply:
SGTM, the numbers looks sorta silly sometimes ("1000 billion cost!")

Reply:
Makes sense to me as well, the error message just reports the factor now.